File: 5-Problem-A-Day/248. Strobogrammatic Number III.java
class Solution {
    char[][] pairs = {{'0', '0'}, {'1', '1'}, {'6', '9'}, {'8', '8'}, {'9', '6'}};

    private int dfs(String low, String high, char[] c, int left, int right) {
        if (left > right) {
            String s = new String(c);
            if ((s.length() == low.length() && s.compareTo(low) < 0)
                    || (s.length() == high.length() && s.compareTo(high) > 0))
                return 0;
            return 1;
        }
        int count = 0;
        for (char[] p : pairs) {
            c[left] = p[0];
            c[right] = p[1];
            if (c.length != 1 && c[0] == '0')
                continue;
            if (left == right && p[0] != p[1])
                continue;
            count += dfs(low, high, c, left + 1, right - 1);
        }
        return count;
    }

    public int strobogrammaticInRange(String low, String high) {
        if (low == null || high == null || low.length() > high.length()
                || (low.length() == high.length() && low.compareTo(high) > 0))
            return 0;
        int count = 0;
        for (int len = low.length(); len <= high.length(); len++)
            count += dfs(low, high, new char[len], 0, len - 1);
        return count;
    }
}

File: 2021/248. Strobogrammatic Number III.cpp
class Solution {
public:
    vector<pair<char, char>> pairs = {{'0', '0'}, {'1', '1'}, {'6', '9'}, {'8', '8'}, {'9', '6'}};
    int count = 0;
    int strobogrammaticInRange(string low, string high) {
        string curr(low.length() - 1, '0');        
        for (int len = low.length(); len <= high.length(); len++) {
            curr += "0";            
            dfs(low, high, curr, 0, len - 1);            
        }
        return count;
    }
    
    void dfs(string& low, string& high, string& curr, int left, int right) {
        if (left > right) {
            if ((curr.length() == low.length() && curr < low)
                || (curr.length() == high.length() && curr > high)) {
                return;
            }
            count++;
            return;
        }
        for (pair<char, char>& p : pairs) {
            curr[left] = p.first;
            curr[right] = p.second;
            if (curr.length() != 1 && curr[0] == '0') continue;
            if (left == right && p.first != p.second) continue;
            dfs(low, high, curr, left + 1, right - 1);
        }
    }
};